### ISSUE TITLE
Python SDK: stale weakref.finalize frees a live Run's registry slot, defeating the double-Run guard

### ISSUE THREAD

**Issue by Srinivasan8888** [evo-hq/evo#97]
### Summary

The Python SDK's active-run registry prevents two `Run` objects with the same `experiment_id` from being live at once. A stale `weakref.finalize` callback can release the slot belonging to a *different, still-active* `Run`, silently defeating that guard.

### Location

`sdk/python/src/evo_agent/_run.py`:
- L82-84 — finalizer armed: `self._finalizer = weakref.finalize(self, _release_active_run, self._experiment_id)`
- L213-214 (`finish`) and L227-228 (`__exit__`) — release the slot with `_ACTIVE_RUNS.discard(self._experiment_id)` but **never** call `self._finalizer.detach()`.

### Root cause

The registry is keyed only by `experiment_id`, not by instance. `finish()`/`__exit__` free the slot but leave the finalizer armed for the object's whole lifetime. When an already-finished `Run` is later garbage-collected, its finalizer fires `_release_active_run(experiment_id)` and discards that id from `_ACTIVE_RUNS` — even though the id now belongs to a **new, still-active** `Run`.

### Impact (empirically reproduced)

```
r1 = Run(experiment_id="x"); r1.finish()      # slot freed
r2 = Run(experiment_id="x")                    # allowed; r2 active, never finished
del r1; gc.collect()                           # r1's stale finalizer runs _release_active_run("x")
# _ACTIVE_RUNS is now empty though r2 is still live
r3 = Run(experiment_id="x")                    # SUCCEEDS — should raise RuntimeError
```

Output:
```
after r1.finish, active: set()
after r2 created (still active), active: {'x'}
after r1 gc'd, active: set() <-- r2 still live!
BUG CONFIRMED: r3 created while r2 still active -> double-Run guard defeated
```

`r2` and `r3` then both run and both try to write `EVO_RESULT_PATH`, colliding at the end — exactly the failure the registry exists to prevent.

### Proposed fix

Detach the finalizer when the slot is released, in both `finish()` and `__exit__`:

```python
with _ACTIVE_RUNS_LOCK:
    _ACTIVE_RUNS.discard(self._experiment_id)
self._finalizer.detach()
```

Optionally also make `_release_active_run` ownership-aware (only discard if this instance still owns the slot) as belt-and-suspenders.

### Verify

Run the snippet above with `PYTHONPATH=sdk/python/src` and a no-op `Backend` subclass (`setup`/`write_trace`/`emit_result`/`emit_gate_summary` as pass).
